include/irr/asset/IVirtualTexture.h

```diff
@@ -0,0 +1,750 @@
+        texData.origsize_x = _img->getCreationParameters().extent.width;
+        texData.origsize_y = _img->getCreationParameters().extent.height;
+
+		texData.pgTab_x = _offset.x;
+		texData.pgTab_y = _offset.y;
+        texData.pgTab_layer = _offset.z;
+
+        //getCreationParameters().mipLevels doesnt necesarilly mean that there wasnt allocated space for higher non-existent mip levels
+        texData.maxMip = _img->getCreationParameters().mipLevels-1u-m_pgSzxy_log2;
+
+        auto ETC_to_int = [](ISampler::E_TEXTURE_CLAMP _etc) -> uint32_t {
+            switch (_etc)
+            {
+            case ISampler::ETC_REPEAT:
+                return STextureData::EWM_REPEAT;
+            case ISampler::ETC_CLAMP_TO_EDGE:
+            case ISampler::ETC_CLAMP_TO_BORDER:
+                return STextureData::EWM_CLAMP;
+            case ISampler::ETC_MIRROR:
+            case ISampler::ETC_MIRROR_CLAMP_TO_EDGE:
+            case ISampler::ETC_MIRROR_CLAMP_TO_BORDER:
+                return STextureData::EWM_MIRROR;
+            default:
+                return STextureData::EWM_INVALID;
+            }
+        };
+
+        texData.wrap_x = ETC_to_int(_wrapu);
+        texData.wrap_y = ETC_to_int(_wrapv);
+
+        return texData;
+    }
+
+    page_tab_offset_t alloc(const IImage* _img, const IImage::SSubresourceRange& _subres, uint32_t _pgtLayer)
+    {
+        const uint32_t pgtAddr2dMask = (1u<<(m_pgtabSzxy_log2*2u))-1u;
+
+        uint32_t szAndAlignment = computeSquareSz(_img, _subres);
+        szAndAlignment *= szAndAlignment;
+
+        uint32_t addr = pg_tab_addr_alctr_t::invalid_address;
+        core::address_allocator_traits<pg_tab_addr_alctr_t>::multi_alloc_addr((*m_pageTableLayerAllocators)[_pgtLayer], 1u, &addr, &szAndAlignment, &szAndAlignment, nullptr);
+        return (addr==pg_tab_addr_alctr_t::invalid_address) ? 
+            page_tab_offset_invalid() :
+            page_tab_offset_t(core::morton2d_decode_x(addr&pgtAddr2dMask), core::morton2d_decode_y(addr&pgtAddr2dMask), _pgtLayer);
+    }
+    virtual bool free(const STextureData& _addr, const IImage* _img, const IImage::SSubresourceRange& _subres)
+    {
+        uint32_t sz = computeSquareSz(_img, _subres);
+        sz *= sz;
+        const uint32_t addr = core::morton2d_encode(_addr.pgTab_x, _addr.pgTab_y);
+
+        core::address_allocator_traits<pg_tab_addr_alctr_t>::multi_free_addr((*m_pageTableLayerAllocators)[_addr.pgTab_layer], 1u, &addr, &sz);
+
+        return true;
+    }
+
+    uint32_t countLevelsTakingAtLeastOnePage(const VkExtent3D& _extent, const ICPUImage::SSubresourceRange& _subres)
+    {
+        const uint32_t baseMaxDim = core::roundUpToPoT(core::max(_extent.width, _extent.height))>>_subres.baseMipLevel;
+        const int32_t lastFullMip = core::findMSB(baseMaxDim) - static_cast<int32_t>(m_pgSzxy_log2);
+
+        return core::clamp(lastFullMip+1, 0, static_cast<int32_t>(m_pageTable->getCreationParameters().mipLevels));
```

Comment:
use only a `max`
assert that `lastFullMip<static_cast<int32_t>(m_pageTable->getCreationParameters().mipLevels)`
because if `lastFullMip` is not within page table mipmap chain then you've tried to pack a texture larger than max allocatable size